vllm/model_executor/models/llava.py

```diff
@@ -0,0 +1,245 @@
+        length dimension.
+        `input_ids` is thus [1, 32000, ..., 32000, 29871, 13, 11889, 29901,
+        1724, 29915, 29879, 278, 2793, 310, 278, 1967, 29973, 13, 22933,
+        9047, 13566, 29901].
+        There will be 576 `32000` in the `input_ids`.
+        (32000 is the token id for `<image>`.)
+
+        This way, the `positions` and `input_metadata` are consistent
+        with the `input_ids`.
+
+        The design is chosen so that all data massaging happens inside
+        Llava model.
+
+        Args:
+            input_ids: Flattened (concatenated) input_ids corresponding to a
+                batch.
+            image_input: A batch of image inputs (either of type PIXEL_VALUES
+                or IMAGE_FEATURES).
+                For PIXEL_VALUES, expecting [1, 3, 336, 336].
+                For IMAGE_FEATURES, expecting [1, 576, 1024].
+        """
+        if image_input is not None:
+            # This should be ensured by ray-llm.
+            # In case vllm entrypoint is used, this is a user facing error.
+            if list(image_input.shape[1:]) != list(
+                    self.vision_language_config.image_input_shape[1:]):
+                raise ValueError(
+                    f"The expected image tensor shape is batch dimension "
+                    f"plus "
+                    f"{self.vision_language_config.image_input_shape[1:]}."
+                    f" You supplied {image_input.shape[1:]}. "
+                    f"If you are using vLLM's entrypoint, make sure your "
+                    f"supplied image input is consistent with "
+                    f"image_input_shape in engine args.")
+            if self.vision_tower is not None:
+                # TODO(xwjiang): Port minimal CLIPVisionModel over.
+                image_outputs = self.vision_tower(image_input,
+                                                  output_hidden_states=True)
+                image_features = image_outputs.hidden_states[
+                    self.config.vision_feature_layer]
+                if self.config.vision_feature_select_strategy == "default":
+                    image_features = image_features[:, 1:]
+            else:
+                image_features = image_input
+            vision_embeddings = self.multi_modal_projector(image_features)
+            inputs_embeds = self.language_model.get_input_embeddings(input_ids)
+            _merge_vision_embeddings(
+                input_ids, inputs_embeds, vision_embeddings,
+                self.vision_language_config.image_token_id)
+
+            hidden_states = self.language_model(None,
+                                                positions,
+                                                kv_caches,
+                                                input_metadata,
+                                                inputs_embeds=inputs_embeds)
+        else:
+            hidden_states = self.language_model(
```

Comment:
I think it will be best to factor this code out of the if condition -- in case of the if condition just set `input_ids = None` and in the else branch set `inputs_embeds = None` and then call
```
hidden_states = self.language_model(
    input_ids,
    positions,
    kv_caches,
    input_metadata,
    inputs_embeds=inputs_embeds
)
```
at the end of the if condition.